llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -121,23 +121,45 @@ static void writeToFile(std::string Filename, std::string Content) {
   OS.close();
 }
 
+// Describes scope covered by each entry in the module-kernel map populated by
+// the function below.
+enum KernelMapEntryScope {
+  Scope_PerKernel, // one entry per kernel
+  Scope_PerModule, // one entry per module
+  Scope_Global     // single entry in the map for all kernels
+};
+
 // Output parameter ResKernelModuleMap is a map containing groups of kernels
 // with same values of the sycl-module-id attribute.
 // The function fills ResKernelModuleMap using input module M.
 static void collectKernelModuleMap(
     Module &M, std::map<StringRef, std::vector<Function *>> &ResKernelModuleMap,
-    bool OneKernelPerModule) {
-
-  constexpr char ATTR_SYCL_MODULE_ID[] = "sycl-module-id";
+    KernelMapEntryScope EntryScope) {
 
   for (auto &F : M.functions()) {
     if (F.getCallingConv() == CallingConv::SPIR_KERNEL) {
-      if (OneKernelPerModule) {
+      switch (EntryScope) {
+      case Scope_PerKernel:
         ResKernelModuleMap[F.getName()].push_back(&F);
-      } else if (F.hasFnAttribute(ATTR_SYCL_MODULE_ID)) {
+        break;
+      case Scope_PerModule: {
+        constexpr char ATTR_SYCL_MODULE_ID[] = "sycl-module-id";
+
+        if (!F.hasFnAttribute(ATTR_SYCL_MODULE_ID))
+          error("no '" + Twine(ATTR_SYCL_MODULE_ID) +
+                "' attribute in kernel '" + F.getName() +
+                "', per-module split not possible");
```

Comment:
Should we group kernels without this attribute into a separate module?

Reply:
I don't know, maybe. The answer would be 'yes' if there are practical scenarios where per-module splitting is requested on modules with functions w/o module-id attribute. This would be a topic for a separate PR, anyway.

Here I just protect from the existing problem that I noticed - silent dropping of kernels w/o module-id attribute in -split=source mode. What I can do now is add a TODO.

Reply:
Sure. I thought about this case while implementing split, but I didn't come up with the viable case when it is possible to get such module containing kernels with and without module-id attribute.